flake8_type_checking/checker.py

```diff
@@ -360,28 +361,147 @@ def import_type(self) -> ImportTypeValue:
         return cast('ImportTypeValue', classify_base(self.full_name.partition('.')[0]))
 
 
-class UnwrappedAnnotation(NamedTuple):
-    """Represents a single `ast.Name` in an unwrapped annotation."""
+class WrappedAnnotation(NamedTuple):
+    """Represents a wrapped annotation i.e. a string constant."""
 
     lineno: int
     col_offset: int
     annotation: str
+    names: set[str]
+    scope: Scope
     type: Literal['annotation', 'alias', 'new-alias']
 
 
-class WrappedAnnotation(NamedTuple):
-    """Represents a wrapped annotation i.e. a string constant."""
+class UnwrappedAnnotation(NamedTuple):
+    """Represents a single `ast.Name` in an unwrapped annotation."""
 
     lineno: int
     col_offset: int
     annotation: str
-    names: set[str]
+    scope: Scope
     type: Literal['annotation', 'alias', 'new-alias']
 
 
+class Symbol(NamedTuple):
+    """Represents an import/definition/declaration of a variable."""
+
+    name: str
+    lineno: int
+    col_offset: int
+    type: Literal['import', 'definition', 'declaration', 'argument']
+    in_type_checking_block: bool
+
+    def available_at_runtime(self, use: HasPosition | None = None) -> bool:
+        """Return whether or not this symbol is available at runtime."""
+        if self.in_type_checking_block or self.type == 'declaration':
+            return False
+
+        # we punt in this case, this is to support some use-cases where
+        # the location of use does not matter, such as __all__
+        if use is None:
+            return True
+
+        if use.lineno < self.lineno:
+            return False
+
+        if use.lineno == self.lineno and use.col_offset < self.col_offset:
+            return False
+
+        return True
+
+
+class Scope:
+    """
+    Represents a scope for looking up symbols.
+
+    We currently don't create a new scope for generator expressions, since it
+    already has a bunch of special cases for accessing symbols that would not
+    be accessible inside a different kind of scope and it may go away entirely
+    when comprehension inlining becomes a thing and it no longer generates a
+    new stack frame.
+
+    For ClassDef/FunctionDef/AsyncFunctionDef we create a tiny virtual scope
+    for the head to properly handle PEP695 parameter scopes.
+    """
+
+    def __init__(self, node: ast.Module | ast.ClassDef | Function, parent: Scope | None = None, is_head: bool = False):
```

Comment:
This is mostly to deal with the fact that parameters need to be available in the function body, but not in the function definition, so in order to not leak the parameters to the outer scope but still avoid ambiguity for stuff like `def foo(date: date) -> None: ...` I needed a virtual scope in-between to properly isolate the symbols. On top of that you now also have type params that need to be available inside the function definition i.e. `def foo[T](x: T) -> T: ...`

With classes it's less obvious and i might be able to get rid of the scope again, since the only problematic case would be `class Foo(Foo):` but since `Foo` is already special-cased in the class body, we shouldn't need an extra scope for isolation.